### PR TITLE
nixos/nginx: enable multiple proxyCachePath support

### PR DESCRIPTION
**Pull request by Izorkin**
###### Description of changes
Issue - https://github.com/NixOS/nixpkgs/issues/222934
Remove option proxyCache and add  multiple proxyCachePath support.

Example configuration:
```
  services.nginx = {
...
    proxyCachePath = {
      "default" = {
        enable = true;
        keysZoneName = "all_cache";
        keysZoneSize = "32m";
        useTempPath = true;
        inactive = "7d";
      };
      "test" = {
        enable = false;
        keysZoneName = "test";
        keysZoneSize = "32m";
        inactive = "14d";
      };
      "local" = {
        enable = true;
        keysZoneName = "CACHE";
        keysZoneSize = "64m";
        levels = "1:2:2";
        inactive = "3d";
      };
    };
...
  };

```

Result:
```
...
  proxy_cache_path /var/cache/nginx/default keys_zone=all_cache:32m levels=1:2 use_temp_path=on inactive=7d max_size=1g;
  proxy_cache_path /var/cache/nginx/local keys_zone=CACHE:64m levels=1:2:2 use_temp_path=off inactive=3d max_size=1g;
...
```

Thanks for your help @RaitoBezarius.

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [x] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: nixos/modules/services/web-servers/nginx/default.nix [hunk 1]
```diff
@@ -240,13 +240,7 @@ let
...
       server_tokens ${if cfg.serverTokens then "on" else "off"};
 
-      ${optionalString cfg.proxyCache.enable ''
-        proxy_cache_path /var/cache/nginx keys_zone=${cfg.proxyCache.keysZoneName}:${cfg.proxyCache.keysZoneSize}
-                                          levels=${cfg.proxyCache.levels}
-                                          use_temp_path=${if cfg.proxyCache.useTempPath then "on" else "off"}
-                                          inactive=${cfg.proxyCache.inactive}
-                                          max_size=${cfg.proxyCache.maxSize};
-      ''}
+      ${mkProxyCache}
```
Comment (RaitoBezarius) [inline nixos/modules/services/web-servers/nginx/default.nix:243]: `mkProxyCache` sounds like a function name, but this is a string here.
Look how `mkLocation` are proper functions.
You should do the same here and call `mkProxyCache something` here.

File: nixos/modules/services/web-servers/nginx/default.nix [hunk 2]
```diff
@@ -434,13 +429,18 @@ let
...
   mkHtpasswd = name: authDef: pkgs.writeText "${name}.htpasswd" (
     concatStringsSep "\n" (mapAttrsToList (user: password: ''
       ${user}:{PLAIN}${password}
     '') authDef)
   );
 
   mkCertOwnershipAssertion = import ../../../security/acme/mk-cert-ownership-assertion.nix;
+
+  mkProxyCache = concatStringsSep "\n" (mapAttrsToList (name: proxyCache: ''
+    proxy_cache_path /var/cache/nginx/${name} keys_zone=${proxyCache.keysZoneName}:${proxyCache.keysZoneSize} levels=${proxyCache.levels} use_temp_path=${if proxyCache.useTempPath then "on" else "off"} inactive=${proxyCache.inactive} max_size=${proxyCache.maxSize};
```
Comment (RaitoBezarius) [inline nixos/modules/services/web-servers/nginx/default.nix:442]: Can you format the parameter on multiple lines so that they are readable?
Also, please format the function calls (`filterAttrs` `mapAttrsToList`, etc.) so we can read them.

File: nixos/modules/services/web-servers/nginx/default.nix [hunk 3]
```diff
@@ -797,7 +797,7 @@ in
       };
 
       proxyCache = mkOption {
-        type = types.submodule {
+        type = types.attrsOf (types.submodule ({ ... }: {
```
Comment (RaitoBezarius) [inline nixos/modules/services/web-servers/nginx/default.nix:800]: This is a breaking change, right? You need to provide some warning / breaking change notice in release notes and in the module itself, when did we introduce `proxyCache`? Do we have a NixOS release with it?
I would argue you should keep `proxyCache` as is and introduce `proxyCaches` for multiple proxy caches.
Deprecate `proxyCache` then remove it in the next release.

**Review by RaitoBezarius — CHANGES_REQUESTED** [review]
Thank you for the change, some stuff relative to breaking changes needs to be done here;

File: nixos/modules/services/web-servers/nginx/default.nix (hunk 3 above)
Comment (Izorkin) [reply nixos/modules/services/web-servers/nginx/default.nix:800]: This option has been recently introduced. I think that few people have used this feature and you can do without a warning.

Comment (RaitoBezarius) [reply nixos/modules/services/web-servers/nginx/default.nix:800]: Unfortunately, a lot of people are on unstable, it just mean we can deprecate it faster.

Comment (Izorkin) [reply nixos/modules/services/web-servers/nginx/default.nix:807]: Renamed optioan and add warning change notice.

File: nixos/modules/services/web-servers/nginx/default.nix [hunk 4]
```diff
@@ -397,6 +402,7 @@ let
         }
       ''
   ) virtualHosts);
+
```
Comment (RaitoBezarius) [inline nixos/modules/services/web-servers/nginx/default.nix:405]: unrelated formatting

File: nixos/modules/services/web-servers/nginx/default.nix [hunk 5]
```diff
@@ -434,6 +440,7 @@ let
     auth_basic secured;
     auth_basic_user_file ${auth_file};
   '');
+
```
Comment (RaitoBezarius) [inline nixos/modules/services/web-servers/nginx/default.nix:443]: unrelated formatting

File: nixos/modules/services/web-servers/nginx/default.nix [hunk 6]
```diff
@@ -102,6 +102,17 @@ let
     proxy_set_header        X-Forwarded-Server $host;
   '';
 
+  cfgProxyCachePath = concatStringsSep "\n" (mapAttrsToList (name: proxyCachePath: ''
```
Comment (RaitoBezarius) [inline nixos/modules/services/web-servers/nginx/default.nix:105]: following the convention, it should be called `proxyCachePathConfig` of this file

File: nixos/doc/manual/release-notes/rl-2305.section.md [hunk 7]
```diff
@@ -284,6 +284,8 @@ In addition to numerous new and upgraded packages, this release has the followin
 
 - Enabling global redirect in `services.nginx.virtualHosts` now allows one to add exceptions with the `locations` option.
 
+- A new option `proxyCachePath` has been added to `services.nginx`. Learn more about proxy_cache_path [here](http://nginx.org/en/docs/http/ngx_http_proxy_module.html#proxy_cache_path).
```
Comment (RaitoBezarius) [inline nixos/doc/manual/release-notes/rl-2305.section.md:287]:
```suggestion
- A new option `proxyCachePath` has been added to `services.nginx`. Learn more about proxy_cache_path: <http://nginx.org/en/docs/http/ngx_http_proxy_module.html#proxy_cache_path>.
```

Comment (RaitoBezarius) [reply nixos/doc/manual/release-notes/rl-2305.section.md:287]: this is a nitpick, not mandatory to apply

File: nixos/modules/services/web-servers/nginx/default.nix [hunk 8]
```diff
@@ -857,9 +864,9 @@ in
               description = lib.mdDoc "Set maximum cache size";
             };
           };
-        };
+        }));
         default = {};
-        description = lib.mdDoc "Configure proxy cache";
+        description = lib.mdDoc "Configure the path and other parameters of a cache";
```
Comment (RaitoBezarius) [inline nixos/modules/services/web-servers/nginx/default.nix:869]:
```suggestion
        description = lib.mdDoc ''
          Configure a proxy cache path entry.
          See <http://nginx.org/en/docs/http/ngx_http_proxy_module.html#proxy_cache_path> for documentation.
       '';
```

File: nixos/modules/services/web-servers/nginx/default.nix [hunk 9]
```diff
@@ -796,10 +803,10 @@ in
           '';
       };
 
-      proxyCache = mkOption {
-        type = types.submodule {
+      proxyCachePath = mkOption {
+        type = types.attrsOf (types.submodule ({ ... }: {
           options = {
-            enable = mkEnableOption (lib.mdDoc "Enable proxy cache");
+            enable = mkEnableOption (lib.mdDoc "Enable configure the path and other parameters of a cache");
```
Comment (RaitoBezarius) [inline nixos/modules/services/web-servers/nginx/default.nix:809]:
```suggestion
            enable = mkEnableOption (lib.mdDoc "this proxy cache path entry");
```

**Review by RaitoBezarius — CHANGES_REQUESTED** [review]
Modulo some formatting changes and docs, I'm mostly happy with this PR.

File: nixos/modules/services/web-servers/nginx/default.nix (hunk 6 above)
Comment (RaitoBezarius) [reply nixos/modules/services/web-servers/nginx/default.nix:105]: This was not solved

File: nixos/modules/services/web-servers/nginx/default.nix [hunk 10]
```diff
@@ -796,10 +801,10 @@ in
           '';
       };
 
-      proxyCache = mkOption {
-        type = types.submodule {
+      proxyCachePath = mkOption {
+        type = types.attrsOf (types.submodule ({ ... }: {
           options = {
-            enable = mkEnableOption (lib.mdDoc "Enable proxy cache");
+            enable = mkEnableOption (lib.mdDoc "This proxy cache path entry");
```
Comment (RaitoBezarius) [inline nixos/modules/services/web-servers/nginx/default.nix:807]: it should be lowercase because `mkEnableOption` will do `Enable *description*`

File: nixos/modules/services/web-servers/nginx/default.nix (hunk 6 above)
Comment (Izorkin) [reply nixos/modules/services/web-servers/nginx/default.nix:105]: Fixed!

File: nixos/modules/services/web-servers/nginx/default.nix (hunk 10 above)
Comment (Izorkin) [reply nixos/modules/services/web-servers/nginx/default.nix:807]: Fixed!

**Review by RaitoBezarius — APPROVED** [review]
Someday, I wish we will have a NixOS test for this feature :P.
Other than this, LGTM.

**Comment by RaitoBezarius** [comment]
@ofborg test nginx

**Comment by Izorkin** [comment]
> Someday, I wish we will have a NixOS test for this feature :P.

I guess it won't be easy. It is necessary to generate a cache, check whether the data is stored in the cache, etc.

**Comment by Izorkin** [comment]
This error occurs in the test for a long time, and I don't know how to fix it.
```
       > webserver #                                  Dload  Upload   Total   Spent    Left  Speed
       > webserver #   0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
       > (finished: must succeed: curl -w '%{http_code}' --head --fail -H 'If-None-Match: "vw879gp4w2cx1z5lnhm6mvrc744g1xr9"' http://localhost/index.html, in 0.03 seconds)
       > Test "check ETag if serving Nix store paths" failed with error: ""
       > cleanup
       > kill machine (pid 8)
       > webserver # qemu-kvm: terminating on signal 15 from pid 6 (/nix/store/syz2y6j53y5hpzbs7l0965zwxshi8iyl-python3-3.10.10/bin/python3.10)
       > (finished: cleanup, in 0.04 seconds)
       > kill vlan (pid 7)
       For full logs, run 'nix log /nix/store/cp2rlj5g6p7dgl53qvkay1qsa76lz4my-vm-test-run-nginx.drv'.
```

**Comment by RaitoBezarius** [comment]
> > Someday, I wish we will have a NixOS test for this feature :P.
> 
> I guess it won't be easy. It is necessary to generate a cache, check whether the data is stored in the cache, etc.

Just asserting that the cache is being written on the disk is already something that tells that NGINX is doing work relative to the cache.
Of course, verifying it does serve the cache is another thing, but it's all about incremental tests. :)

**Comment by RaitoBezarius** [comment]
Some editorconfig indentation stuff left to do. :)

**Comment by Izorkin** [comment]
> Some editorconfig indentation stuff left to do. :)

Do I need to change something in the configuration?

**Comment by RaitoBezarius** [comment]
Look to the CI: https://github.com/NixOS/nixpkgs/actions/runs/4525800585/jobs/7970537176?pr=223085 -- there's some ident problems in the file you touched. I guess tabs or whatever got mixed.

**Comment by Izorkin** [comment]
Fixed.

**Comment by RaitoBezarius** [comment]
@ofborg eval

**Review by RaitoBezarius — APPROVED** [review]

File: nixos/doc/manual/release-notes/rl-2305.section.md [hunk 11]
```diff
@@ -296,6 +296,8 @@ In addition to numerous new and upgraded packages, this release has the followin
 
 - Enabling global redirect in `services.nginx.virtualHosts` now allows one to add exceptions with the `locations` option.
 
+- A new option `proxyCachePath` has been added to `services.nginx`. Learn more about proxy_cache_path: <http://nginx.org/en/docs/http/ngx_http_proxy_module.html#proxy_cache_path>.
```
Comment (RaitoBezarius) [inline nixos/doc/manual/release-notes/rl-2305.section.md:299]:
```suggestion
- A new option `proxyCachePath` has been added to `services.nginx`. Learn more about proxy_cache_path: <https://nginx.org/en/docs/http/ngx_http_proxy_module.html#proxy_cache_path>.
```

(notice the `s` :P)

**Review by SuperSandro2000 — APPROVED** [review]

File: nixos/modules/services/web-servers/nginx/default.nix [hunk 12]
```diff
@@ -102,6 +102,17 @@ let
     proxy_set_header        X-Forwarded-Server $host;
   '';
 
+  proxyCachePathConfig = concatStringsSep "\n" (mapAttrsToList (name: proxyCachePath: ''
+    proxy_cache_path ${concatStringsSep " " [
+      "/var/cache/nginx/${name}"
```
[the remark this answers is not in the document]
Comment (Izorkin) [reply nixos/modules/services/web-servers/nginx/default.nix:107]: Change to `"/var/cache/nginx"` ?
This variant not working:
```
nginx: [emerg] the same path name "/var/cache/nginx" used in /etc/nginx/nginx.conf:32 and in /etc/nginx/nginx.conf:33
```

File: nixos/doc/manual/release-notes/rl-2305.section.md (hunk 11 above)
Comment (Izorkin) [reply nixos/doc/manual/release-notes/rl-2305.section.md]: Fixed!

File: nixos/modules/services/web-servers/nginx/default.nix [hunk 13]
```diff
@@ -240,14 +253,6 @@ let
 
       server_tokens ${if cfg.serverTokens then "on" else "off"};
 
-      ${optionalString cfg.proxyCache.enable ''
```
[the remark this answers is not in the document]
Comment (Izorkin) [reply nixos/modules/services/web-servers/nginx/default.nix:243]: There is no reason.
Lowered below.

**Review by SuperSandro2000 — APPROVED** [review]

**Review by fpletz — APPROVED** [review]
